src/hashset.c

```diff
@@ -1651,24 +1673,32 @@ void hashsetDump(hashset *s) {
 }
 
 void hashsetHistogram(hashset *s) {
-    for (int table = 0; table <= 1; table++) {
-        for (size_t idx = 0; idx < numBuckets(s->bucket_exp[table]); idx++) {
-            bucket *b = &s->tables[table][idx];
+    for (int table = 0; table <= 1 && s->bucket_exp[table] >= 0; table++) {
+        size_t mask = expToMask(s->bucket_exp[table]);
+        size_t cursor = 0;
+        do {
+            bucket *b = &s->tables[table][cursor];
             char c = b->presence == 0 && b->everfull ? 'X' : '0' + __builtin_popcount(b->presence);
             printf("%c", c);
-        }
+
+            cursor = nextCursor(cursor, mask);
+        } while (cursor != 0);
         if (table == 0) printf(" ");
     }
     printf("\n");
 }
 
 void hashsetProbeMap(hashset *s) {
-    for (int table = 0; table <= 1; table++) {
-        for (size_t idx = 0; idx < numBuckets(s->bucket_exp[table]); idx++) {
-            bucket *b = &s->tables[table][idx];
+    for (int table = 0; table <= 1 && s->bucket_exp[table] >= 0; table++) {
+        size_t mask = expToMask(s->bucket_exp[table]);
+        size_t cursor = 0;
+        do {
+            bucket *b = &s->tables[table][cursor];
             char c = b->everfull ? 'X' : 'o';
             printf("%c", c);
-        }
+
+            cursor = nextCursor(cursor, mask);
+        } while (cursor != 0);
```

Comment:
Only debug print functions for human eyes. I prefer to see it in rehash/scan/iteration/probe order.